test-integration/reference-book.coffee

```diff
@@ -12,65 +13,91 @@ describe 'Reference Book Exercises', ->
   @it 'Loads Biology reference book (readonly)', ->
     @login(TEACHER_USERNAME)
 
-
-    checkForMissingExercises = =>
+    checkSectionsForMissingExercises = =>
       # Wait until the book has loaded.
-      @addTimeout(60)
       @waitAnd(css: '.page-wrapper .page.has-html')
+        .then(closeMenuAndResizeBook)
+        .then =>
+          # Perform `checkEachPage` in series for SECTIONS_TO_TEST times.
+          @forNTimesInSeries(SECTIONS_TO_TEST, checkEachPage)()
+
+    closeMenuAndResizeBook = =>
+      # Resize prevents need to click next again -- without this,
+      # if the window is too skinny/if the next button is right by the scroll bar,
+      # the page will accidentally get clicked to scroll as opposed to clicking next.
+      @driver.manage().window().setSize(1080, 1080)
       @waitClick(css: '.menu-toggle')
 
-      doneLoading = =>
-        # Wait until the modal closes after clicking the date
-        @driver.isElementPresent(css: '.loadable.is-loading, .loading-exercise').then (isPresent) -> not isPresent
-
-      for i in [1..SECTIONS_TO_TEST]
-        # Selenium sometimes keeps pressing the same next button (doneLoading doesn't seem to work 100%)
-        @driver.findElement(css: 'a.nav.next').getAttribute('href').then (oldHref) =>
-          @addTimeout(3)
+    checkEachPage = =>
+      hrefToCheck = null
+      nextNav = null
+      # Selenium sometimes keeps pressing the same next button (doneLoading doesn't seem to work 100%)
+      @driver.wait(doneLoading)
+        .then =>
```

Comment:
@philschatz points out that selenium's promises que up before running in series: https://code.google.com/p/selenium/wiki/WebDriverJs#Control_Flows